internal/servicelog/formatter.go

```diff
@@ -151,20 +158,70 @@ func (w *TrimWriter) Write(p []byte) (int, error) {
 	}
 }
 
-func (w *TrimWriter) write(p []byte) error {
+func (w *TrimWriter) write(line []byte) error {
+	const maxDetectFailures = 10
+	if w.layout == "" && w.detectFailures < maxDetectFailures {
+		w.layout = detectLayout(line)
+		if w.layout == "" {
+			w.detectFailures++
+		}
+	}
+
 	start := 0
-	loc := w.re.FindIndex(p)
-	if loc != nil {
-		start = loc[1]
+	if w.layout != "" {
+		prefix := timeLayoutPrefix(w.layout, line)
+
+		// TODO: do we want to just trust the field splitting and not try to
+		// parse every prefix?  This could potentially cause us to discard
+		// non-timestamp content from log lines.
+		_, err := time.Parse(w.layout, prefix)
+		if err == nil {
+			start = len(prefix)
+		} else {
+			// TODO: should this be an error?
+		}
 	}
 
-	p = p[start:]
-	for len(p) > 0 {
-		n, err := w.dest.Write(p)
-		p = p[n:]
+	line = line[start:]
+	for len(line) > 0 {
+		n, err := w.dest.Write(line)
+		line = line[n:]
 		if err != nil {
 			return err
 		}
 	}
 	return nil
 }
+
+func timeLayoutPrefix(layout string, line []byte) string {
+	if layout == "" {
+		return ""
+	}
+
+	nfields := len(strings.Split(layout, " "))
+	sep := []byte{' '}
+	fields := bytes.SplitN(line, sep, nfields)
```

Comment:
Is this enough, or do we actually have to use `bytes.Fields` (or a similar, custom `FieldsN` type of function) so that runs of whitespace get coalesced into one? For example, times in `time.UnixDate` format may have two spaces before the day number, like `Mon Jan  2 15:04:05 MST 2006`. I'm guessing this code would fail for that?

Reply:
I thought about this a bit, but doesn't the UnixDate format still only have one space?  The `_` in the layout string just indicates that a zero prefix to the digit is omitted.  I was under the impression that it wasn't possible to have multi-space runs that aren't explicitly present in the layout string.  I did it this way because it's much simpler to determine the length of the prefix - using `Fields` makes a bit more onerous because it may gobble up runs of whitespace.

Reply:
No, I think the `_` indicates an additional space if needed. See example here: https://go.dev/play/p/DIF6Deq8C8X and from the `time` package docs:

> Within the format string, the underscores in "_2" and "__2" represent spaces that may be replaced by digits if the following number has multiple digits, for compatibility with fixed-width Unix time formats. A leading zero represents a zero-padded value.
>
> The formats __2 and 002 are space-padded and zero-padded three-character day of year; there is no unpadded day of year format.

Reply:
This has been addressed with the latest changes.